File: git-remove-stale.js
var child_process = require('child_process');
var async = require('async');

var argv = require('minimist')(process.argv, {
  string: 'remote',
  boolean: ['do-it', 'force'],
  'default': {
    'remote': 'origin',
    'force': false
  }
});

var options = ['do-it', 'force', 'remote', '_'];
var validParams = Object.keys(argv).some(function (name) {
  if (options.indexOf(name) == -1) {
    console.info('Usage: git-remove-stale --do-it --force --remote {remote}');
    return true;
  }
});

// Split the stdout and stderr output
// and will take out all the empty lines
function asyncSplit() {
  return function (stdout, stderr, callback) {
    var lines = stdout.split('\n').map(function (line) {
      return line.trim();
      })
      // remove empty
      .filter(function (line) {
        return line != '';
      });

    return callback(null, lines);
  };
}

function asyncExec(argsArray, skipError) {
  return function (callback) {
    child_process.exec(argsArray.join(' '), function (err, stdout, stderr) {
      if (err) {
        if (skipError) {
          return callback(null, stdout, stderr);
        }

        return callback(err);
      }

      return callback(null, stdout, stderr);
    });
  }
};

var obj = {
  run: function (callback) {

    // cached braches from the remote
    this.remoteBranches = [];

    // local branches which are checkout from the remote
    this.localBranches = [];

    // branches which are available locally but not remotly
    this.staleBranches = [];

    // branches which are available on host
    this.liveBranches = [];

    async.waterfall([
      this.findLocalBranches.bind(this),
      this.findRemoteBranches.bind(this),
      this.findLiveBranches.bind(this),
      this.analyzeLiveAndCache.bind(this),
      this.findStaleBranches.bind(this),
      this.deleteBranches.bind(this)
    ], callback);

  },



  _getRemoteForBranch: function (branchName, callback) {
    var exec = asyncExec([
      'git', 'config',
      '--get', 'branch.%s.remote'.replace('%s', branchName)]
    );

    exec(function (err, stdout, stderr) {
      if (err) {
        return callback(null, {
          name: branchName,
          remote: ''
        });
      }

      return callback(null, {
        name: branchName,
        remote: stdout.trim()
      });
    });
  },

  findLocalBranches: function(callback) {
    var _this = this;

    async.waterfall([

      // list all the branches
      asyncExec(['git', 'branch']),

      asyncSplit(),

      // take out star if active branch
      function (lines, h) {
        var branches = lines.map(function (branchName) {
          return branchName.replace(/\*/g, '').trim();
        });

        return h(null, branches);
      },

      // get remote for every branch
      function (branches, h) {
        async.map(branches, this._getRemoteForBranch.bind(this), h);
      }.bind(this),

      // filter branches based on remote
      function (branches, h) {
        branches = branches.filter(function (branch) {
          return branch.remote == argv.remote;
        });

        return h(null, branches);
      }

    ], function (err, branches) {

      _this.localBranches = branches.map(function (branch) {
        return branch.name;
      });

      return callback();
    });

  },

  //
  // this method will use "git ls-remote"
  // to find branches which are still available on the remote
  // and store them in liveBranches state
  //
  findLiveBranches: function (callback) {
    var _this = this;

    async.waterfall([
      // get list of remote branches from remote host
      asyncExec(['git', 'ls-remote', '-h', argv.remote]),

      asyncSplit(),

      // take out sha and refs/heads
      function (lines, h) {
        var correct_branches = [];
        lines.forEach(function (line) {
          var group = line.match(/refs\/heads\/([^\s]*)/);
          if (group) {
            correct_branches.push(group[1]);
          }
        });

        return h(null, correct_branches);
      }

    ], function (err, lines) {

      if (err) {

        _this.liveBranches = [];

        if (err.code && err.code == '128') {
          // error 128 means there is no connection currently to the remote
          // skip this step then
          _this.noConnection = true;
          return callback(null);
        } 

        return callback(err);
      }

      _this.liveBranches = lines;

      return callback(null);
    });
  },

  findRemoteBranches: function(callback) {
    var _this = this;

    async.waterfall([

      // get list of remote branches
      asyncExec(['git', 'branch', '-r']),

      //split lines
      asyncSplit(),

      // filter out non origin branches
      function (branches, h) {
        correct_branches = [];
        var re = new RegExp('^%s\\/([^\\s]*)'.replace('%s', argv.remote));
        branches.forEach(function (branchName) {
          var group = branchName.match(re);
          if (group) {
            correct_branches.push(group[1]);
          }
        });

        return h(null, correct_branches);
      }

    ], function (err, branches) {
      _this.remoteBranches = branches;
      return callback(null);
    });

  },

  //
  // this method will look which branches on remote are absent
  // but still available in here in remotes
  //
  analyzeLiveAndCache: function (callback) {
    if (this.noConnection) {
      // unable to determinate remote branches, because host is not available
      console.warn('WARNING: Unable to connect to remote host');
      return callback();
    } else {
      var message = [
        'WARNING: Your git repository is outdated, please run "git fetch -p"',
        '         Following branches are not pruned:',
        ''
      ];
      var toRemove = [];
      var show = false;

      // compare absent remotes
      this.remoteBranches.forEach(function (branch) {
        if (branch == 'HEAD') {
        } else if (this.liveBranches.indexOf(branch) == -1) {
          message.push('         - ' + branch);
          show = true;
        }
      }, this);

      message.push('');

      if (show) {
        console.warn(message.join('\r\n'));
      }

      this.remoteBranches = this.liveBranches;
      
      return callback();
    }
  },

  findStaleBranches: function(callback) {

    this.localBranches.forEach(function (localBranch) {
      if (this.remoteBranches.indexOf(localBranch) == -1) {
        this.staleBranches.push(localBranch);
      }
    }, this);

    return callback(null);
  },

  deleteBranches: function(callback) {

    if (!this.staleBranches.length) {
      console.info('No stale branches are found');
      return callback();
    }

    async.forEach(this.staleBranches, function (branchName, h) {
      if (argv['do-it']) {
        console.info('Removing "' + branchName + '"');

        var dFlag  = argv.force ? '-D' : '-d';
        var exec = asyncExec(['git', 'branch', dFlag, branchName]);
        exec(function (err, stdout, stderr) {
          if (err) {
            console.error('ERROR: Unable to remove: ' + err.message);
            return h();
          }

          console.info(stdout);
          return h();
        });
      } else {
        console.info('Found branch "' + branchName + '"');
        return h();
      }
    }, function (err) {
      console.info('');
      console.info('INFO: To remove found branches use --do-it flag');
      return callback();
    });

  }
}


if (!validParams) {
  // check for git repository
  var exec = asyncExec(['git', 'rev-parse', '--show-toplevel']);
  
  exec(function (err, stdout, stderr) {
    if (err) {
      console.error(err.message);
      return;
    }

    obj.run();
  });
}

File: git-remove-stale.py
#!/usr/bin/env python

import subprocess
import re
import argparse

parser = argparse.ArgumentParser(description="Remove local branches, which are no longer available in the remote")
parser.add_argument("--do-it", action="store_true", help="Remove branches")
parser.add_argument("--force", action="store_true", help="Force deletion")
parser.add_argument("--remote", default="origin", help="Remote name")
args = parser.parse_args()



def find_local_branches():
  branches = subprocess.check_output(["git", "branch"])
  correct_branches = [];
  for line in branches.splitlines():
    # take out the active branch, we are only intersted in the name of the
    # branch
    branch_name = re.sub(r"\*", "", line).strip()

    # find out what is the remote of the branch
    try:
      remoteName = subprocess.check_output(["git", "config", "--get", "branch.%s.remote" % branch_name])
    except Exception as e:
      # Branch has no config"
      remoteName = ""

    remoteName = remoteName.strip();

    if "%s" % remoteName == args.remote:
      correct_branches.append(branch_name)

  return correct_branches


def find_remote_branches():
  branches = subprocess.check_output(["git", "branch", "-r"])
  correct_branches = [];

  for line in branches.splitlines():
    branch_name = line.strip();
    result = re.search(r"%s/([^\s]*)" % args.remote, branch_name)
    if result:
      correct_branches.append(result.group(1))

  return correct_branches

def find_live_branches():
  try:
    branches = subprocess.check_output(["git", "ls-remote", "-h", args.remote])
  except Exception as e:
    #TODO: test that this is error 128

    return None

  for line in branches.splitlines():
    branch_name = line.strip();
    result = re.search(r"/refs/heads/([^\s]*)", branch_name)

    if result:
      correct_branches.append(result.group(1))

  return correct_branches


def remove_branches(branches):
  for branch_name in branches:
    if (args.do_it):
      print "Removing %s" % branch_name
      if argv.force:
        deleteFlag = "-D"
      else:
        deleteFlag = "-d"

      return_code = subprocess.call(["git", "branch", deleteFlag, branch_name]);
      if return_code != 0:
        print "Unable to remove branch"
    else:
      print "Will remove %s" % branch_name

def analyze_live_and_remote(live_branches, remote_branches):
  if live_branches == None:
    return remote_branches

  notFound = []

  for branch_name in remote_branches:
    try:
      index = live_branches.index(branch_name)
    except ValueError:
      notFound.append(branch_name)

  if notFound:
    print '''
      WARNING: Your repository is not up-to-date with remote
    '''

  return live_branches

def find_to_remove(local=[], remote=[]):
  will_remove = []

  for branch_name in local_branches:
    try:
      index = remote_branches.index(branch_name)
    except ValueError:
      will_remove.append(branch_name)

  return will_remove



# test whether this is the git repo
try:
  subprocess.check_output(["git", "rev-parse", "--show-toplevel"]);
  is_git = True
except Exception:
  is_git = False

if is_git:
  # walk through the local branches
  # if local branch is not in remote branch list
  # prepare for removing
  remote_branches = find_remote_branches()
  local_branches = find_local_branches() 
  live_branches = find_live_branches()
  remote_branches = analyze_live_and_remote(live_branches, remote_branches)

  to_remove = find_to_remove(local=local_branches, remote=remote_branches)

  if to_remove:
    remove_branches(to_remove)
  else:
    print "No stale branches are found"
